{
	"schema_version": "1.6.0",
	"id": "CVE-2021-4044",
	"modified": "2023-12-06T01:01:31.831252Z",
	"published": "2021-12-14T19:15:07Z",
	"aliases": [
		"BIT-node-2021-4044",
		"GHSA-mmjf-f5jw-w72q",
		"RUSTSEC-2021-0129"
	],
	"details": "Internally libssl in OpenSSL calls X509_verify_cert() on the client side to verify a certificate supplied by a server. That function may return a negative return value to indicate an internal error (for example out of memory). Such a negative return value is mishandled by OpenSSL and will cause an IO function (such as SSL_connect() or SSL_do_handshake()) to not indicate success and a subsequent call to SSL_get_error() to return the value SSL_ERROR_WANT_RETRY_VERIFY. This return value is only supposed to be returned by OpenSSL if the application has previously called SSL_CTX_set_cert_verify_callback(). Since most applications do not do this the SSL_ERROR_WANT_RETRY_VERIFY return value from SSL_get_error() will be totally unexpected and applications may not behave correctly as a result. The exact behaviour will depend on the application but it could result in crashes, infinite loops or other similar incorrect responses. This issue is made more serious in combination with a separate bug in OpenSSL 3.0 that will cause X509_verify_cert() to indicate an internal error when processing a certificate chain. This will occur where a certificate does not include the Subject Alternative Name extension but where a Certificate Authority has enforced name constraints. This issue can occur even with valid chains. By combining the two issues an attacker could induce incorrect, application dependent behaviour. Fixed in OpenSSL 3.0.1 (Affected 3.0.0).",
	"severity": [
		{
			"type": "CVSS_V3",
			"score": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H"
		}
	],
	"affected": [
		{
			"package": {
				"ecosystem": "Alpine:v3.15",
				"name": "openssl3",
				"purl": "pkg:apk/alpine/openssl3?arch=source"
			},
			"ranges": [
				{
					"type": "ECOSYSTEM",
					"events": [
						{
							"introduced": "0"
						},
						{
							"fixed": "3.0.2-r0"
						}
					]
				}
			],
			"database_specific": {
				"source": "https://storage.googleapis.com/cve-osv-conversion/osv-output/CVE-2021-4044.json"
			}
		},
		{
			"package": {
				"ecosystem": "Alpine:v3.16",
				"name": "openssl3",
				"purl": "pkg:apk/alpine/openssl3?arch=source"
			},
			"ranges": [
				{
					"type": "ECOSYSTEM",
					"events": [
						{
							"introduced": "0"
						},
						{
							"fixed": "3.0.1-r0"
						}
					]
				}
			],
			"database_specific": {
				"source": "https://storage.googleapis.com/cve-osv-conversion/osv-output/CVE-2021-4044.json"
			}
		},
		{
			"package": {
				"ecosystem": "Alpine:v3.17",
				"name": "openssl",
				"purl": "pkg:apk/alpine/openssl?arch=source"
			},
			"ranges": [
				{
					"type": "ECOSYSTEM",
					"events": [
						{
							"introduced": "0"
						},
						{
							"fixed": "3.0.1-r0"
						}
					]
				}
			],
			"database_specific": {
				"source": "https://storage.googleapis.com/cve-osv-conversion/osv-output/CVE-2021-4044.json"
			}
		},
		{
			"package": {
				"ecosystem": "Alpine:v3.18",
				"name": "openssl",
				"purl": "pkg:apk/alpine/openssl?arch=source"
			},
			"ranges": [
				{
					"type": "ECOSYSTEM",
					"events": [
						{
							"introduced": "0"
						},
						{
							"fixed": "3.0.1-r0"
						}
					]
				}
			],
			"database_specific": {
				"source": "https://storage.googleapis.com/cve-osv-conversion/osv-output/CVE-2021-4044.json"
			}
		},
		{
			"package": {},
			"ranges": [
				{
					"type": "GIT",
					"repo": "https://github.com/nodejs/node",
					"events": [
						{
							"fixed": "b65a11e072cf751d8c2a9699451ef80022d4ffbc"
						},
						{
							"introduced": "f99ce7c1b88ff445f60e9e5575b674cb509e47f3"
						}
					]
				},
				{
					"type": "GIT",
					"repo": "https://github.com/openssl/openssl",
					"events": [
						{
							"introduced": "0"
						},
						{
							"fixed": "e818b74be2170fbe957a07b0da4401c2b694b3b8"
						}
					]
				}
			],
			"versions": [
				"BEFORE_engine",
				"OpenSSL_0_9_1c",
				"OpenSSL_0_9_2b",
				"OpenSSL_0_9_3",
				"OpenSSL_0_9_3a",
				"OpenSSL_0_9_3beta2",
				"OpenSSL_0_9_4",
				"OpenSSL_0_9_5a",
				"OpenSSL_0_9_5a-beta1",
				"OpenSSL_0_9_5a-beta2",
				"OpenSSL_0_9_5beta1",
				"OpenSSL_0_9_5beta2",
				"OpenSSL_0_9_6-beta3",
				"OpenSSL_1_0_0",
				"OpenSSL_1_0_0-beta1",
				"OpenSSL_1_0_0-beta2",
				"OpenSSL_1_0_0-beta3",
				"OpenSSL_1_0_0-beta4",
				"OpenSSL_1_0_0-beta5",
				"OpenSSL_1_0_0a",
				"OpenSSL_1_0_1",
				"OpenSSL_1_0_1-beta1",
				"OpenSSL_1_0_1-beta2",
				"OpenSSL_1_0_1-beta3",
				"OpenSSL_1_0_2",
				"OpenSSL_1_0_2-beta1",
				"OpenSSL_1_0_2-beta2",
				"OpenSSL_1_0_2-beta3",
				"OpenSSL_1_0_2-post-auto-reformat",
				"OpenSSL_1_0_2-post-reformat",
				"OpenSSL_1_0_2-pre-auto-reformat",
				"OpenSSL_1_0_2-pre-reformat",
				"OpenSSL_1_0_2a",
				"OpenSSL_1_0_2b",
				"OpenSSL_1_0_2c",
				"OpenSSL_1_0_2d",
				"OpenSSL_1_0_2e",
				"OpenSSL_1_0_2f",
				"OpenSSL_1_0_2g",
				"OpenSSL_1_0_2h",
				"OpenSSL_1_0_2i",
				"OpenSSL_1_0_2j",
				"OpenSSL_1_0_2k",
				"OpenSSL_1_0_2l",
				"OpenSSL_1_0_2m",
				"OpenSSL_1_0_2n",
				"OpenSSL_1_0_2o",
				"OpenSSL_1_0_2p",
				"OpenSSL_1_0_2q",
				"OpenSSL_1_0_2r",
				"OpenSSL_1_0_2s",
				"OpenSSL_1_0_2t",
				"v17.0.0"
			],
			"database_specific": {
				"source": "https://storage.googleapis.com/cve-osv-conversion/osv-output/CVE-2021-4044.json"
			}
		}
	],
	"references": [
		{
			"type": "ADVISORY",
			"url": "https://security.netapp.com/advisory/ntap-20211229-0003/"
		},
		{
			"type": "ADVISORY",
			"url": "https://www.openssl.org/news/secadv/20211214.txt"
		},
		{
			"type": "WEB",
			"url": "https://git.openssl.org/gitweb/?p=openssl.git%3Ba=commitdiff%3Bh=758754966791c537ea95241438454aa86f91f256"
		}
	]
}
